File: golang/02-never-ending-function/order_processor_test.go
package kata

import (
	"strings"
	"testing"
)

func TestProcessOrderRejectsEmptyItems(t *testing.T) {
	op := NewOrderProcessor()
	order := Order{Items: []Item{}, Customer: Customer{Email: "a@b.com"}, Address: Address{Zip: "12345"}}
	result := op.process_order(order)
	if result.Error != "No items" {
		t.Errorf("expected error 'No items', got %q", result.Error)
	}
}

func TestProcessOrderRejectsInvalidCustomer(t *testing.T) {
	op := NewOrderProcessor()
	order := Order{Items: []Item{{Price: 10, Quantity: 1}}, Customer: Customer{}, Address: Address{Zip: "12345"}}
	result := op.process_order(order)
	if result.Error != "Invalid customer" {
		t.Errorf("expected error 'Invalid customer', got %q", result.Error)
	}
}

func TestProcessOrderRejectsInvalidAddress(t *testing.T) {
	op := NewOrderProcessor()
	order := Order{Items: []Item{{Price: 10, Quantity: 1}}, Customer: Customer{Email: "a@b.com"}, Address: Address{}}
	result := op.process_order(order)
	if result.Error != "Invalid address" {
		t.Errorf("expected error 'Invalid address', got %q", result.Error)
	}
}

func TestProcessOrderRejectsOutOfStock(t *testing.T) {
	op := NewOrderProcessor()
	order := Order{Items: []Item{{Price: 10, Quantity: 101}}, Customer: Customer{Email: "a@b.com"}, Address: Address{Zip: "12345"}}
	result := op.process_order(order)
	if result.Error != "Out of stock" {
		t.Errorf("expected error 'Out of stock', got %q", result.Error)
	}
}

func TestProcessOrderCalculatesTotalsForSmallOrder(t *testing.T) {
	op := NewOrderProcessor()
	order := Order{Items: []Item{{Price: 10, Quantity: 1}}, Customer: Customer{Email: "a@b.com"}, Address: Address{Zip: "12345"}}
	result := op.process_order(order)
	if result.Total != 16.69 {
		t.Errorf("expected total 16.69, got %v", result.Total)
	}
	if result.Carrier != "USPS" {
		t.Errorf("expected carrier USPS, got %q", result.Carrier)
	}
}

func TestProcessOrderCalculatesTotalsForMediumOrder(t *testing.T) {
	op := NewOrderProcessor()
	order := Order{Items: []Item{{Price: 20, Quantity: 1}}, Customer: Customer{Email: "a@b.com"}, Address: Address{Zip: "12345"}}
	result := op.process_order(order)
	if result.Total != 27.39 {
		t.Errorf("expected total 27.39, got %v", result.Total)
	}
}

func TestProcessOrderUsesUPSForLargeOrders(t *testing.T) {
	op := NewOrderProcessor()
	order := Order{Items: []Item{{Price: 100, Quantity: 1}}, Customer: Customer{Email: "a@b.com"}, Address: Address{Zip: "12345"}}
	result := op.process_order(order)
	if result.Carrier != "UPS" {
		t.Errorf("expected carrier UPS, got %q", result.Carrier)
	}
	if result.PaymentStatus != "approved" {
		t.Errorf("expected payment status 'approved', got %q", result.PaymentStatus)
	}
}

func TestProcessOrderFlagsHighTotalAsPendingReview(t *testing.T) {
	op := NewOrderProcessor()
	order := Order{Items: []Item{{Price: 1000, Quantity: 1}}, Customer: Customer{Email: "a@b.com"}, Address: Address{Zip: "12345"}}
	result := op.process_order(order)
	if result.PaymentStatus != "pending_review" {
		t.Errorf("expected payment status 'pending_review', got %q", result.PaymentStatus)
	}
}

func TestProcessOrderIncludesEmailConfirmation(t *testing.T) {
	op := NewOrderProcessor()
	order := Order{Items: []Item{{Price: 10, Quantity: 1}}, Customer: Customer{Email: "user@test.com"}, Address: Address{Zip: "12345"}}
	result := op.process_order(order)
	if result.EmailTo != "user@test.com" {
		t.Errorf("expected email to 'user@test.com', got %q", result.EmailTo)
	}
	if !strings.HasPrefix(result.OrderID, "ORD-") {
		t.Errorf("expected OrderID to start with 'ORD-', got %q", result.OrderID)
	}
}

File: csharp/02-never-ending-function/test/order_processor_test.cs
using Xunit;

public class OrderProcessorTest
{
    private static Order CreateOrder(decimal price = 10, int quantity = 1, string email = "a@b.com", string zip = "12345")
    {
        return new Order(
            new List<LineItem> { new LineItem(price, quantity) },
            new Customer(email),
            new Address(zip)
        );
    }

    [Fact]
    public void RejectsEmptyItems()
    {
        var processor = new OrderProcessor();
        var result = processor.process_order(new Order(new List<LineItem>(), new Customer("a@b.com"), new Address("12345")));
        Assert.Equal("No items", result.Error);
    }

    [Fact]
    public void RejectsInvalidCustomer()
    {
        var processor = new OrderProcessor();
        var result = processor.process_order(CreateOrder(email: ""));
        Assert.Equal("Invalid customer", result.Error);
    }

    [Fact]
    public void RejectsInvalidAddress()
    {
        var processor = new OrderProcessor();
        var result = processor.process_order(CreateOrder(zip: ""));
        Assert.Equal("Invalid address", result.Error);
    }

    [Fact]
    public void RejectsOutOfStock()
    {
        var processor = new OrderProcessor();
        var result = processor.process_order(CreateOrder(quantity: 101));
        Assert.Equal("Out of stock", result.Error);
    }

    [Fact]
    public void CalculatesTotalsForSmallOrder()
    {
        var processor = new OrderProcessor();
        var result = processor.process_order(CreateOrder());
        Assert.Equal(16.69m, result.Total);
        Assert.Equal("USPS", result.Carrier);
    }

    [Fact]
    public void CalculatesTotalsForMediumOrder()
    {
        var processor = new OrderProcessor();
        var result = processor.process_order(CreateOrder(price: 20));
        Assert.Equal(27.39m, result.Total);
    }

    [Fact]
    public void UsesUpsForLargeOrders()
    {
        var processor = new OrderProcessor();
        var result = processor.process_order(CreateOrder(price: 100));
        Assert.Equal("UPS", result.Carrier);
        Assert.Equal("approved", result.PaymentStatus);
    }

    [Fact]
    public void FlagsHighTotalAsPendingReview()
    {
        var processor = new OrderProcessor();
        var result = processor.process_order(CreateOrder(price: 1000));
        Assert.Equal("pending_review", result.PaymentStatus);
    }

    [Fact]
    public void IncludesEmailConfirmation()
    {
        var processor = new OrderProcessor();
        var result = processor.process_order(CreateOrder(email: "user@test.com"));
        Assert.Equal("user@test.com", result.EmailTo);
        Assert.Equal("Order Confirmation", result.EmailSubject);
    }

    [Fact]
    public void Quantity100NotOutOfStock()
    {
        var processor = new OrderProcessor();
        var result = processor.process_order(CreateOrder(price: 1, quantity: 100));
        Assert.Null(result.Error);
    }
}
